### PR TITLE
Add Zero-Copy Storage Sharing from CPU to GPU for Fast Model Loading under Unified Memory Architecture Platforms

### PR DESCRIPTION
**Pull request by nagic0**
## Summary

This PR represents a first attempt to introduce Unified Shared Memory (USM) or Unified Memory Architecture(UMA) capabilities to PyTorch.

While achieving a completely zero-copy execution pipeline (where all tensor operations utilize shared memory by default) involves complex architectural changes (e.g. #72581, #107605, #145693, and https://github.com/pytorch/rfcs/pull/36), this PR focuses on a part of it: **optimizing the model loading phase**.

It introduces mechanisms to share CPU-resident storage directly with the GPU without data copy, significantly accelerating `from_file` loading and reducing model cold-start time on UMA platforms like Intel iGPU, Apple Silicon, NVIDIA Jetson, and AMD APU. Benchmarks demonstrate a **35% to 80% reduction in model loading latency** on UMA platforms.

## Motivation & Scope

On UMA platforms, physically shared RAM allows for zero-copy access, but PyTorch's current default behavior involves redundant copies:

1. Disk to OS Page Cache (with mmap)
2. Explicit Copy from Page Cache to GPU Memory

Implementing a global zero-copy allocator for all PyTorch operations is a massive undertaking. Therefore, this PR is scoped specifically to Storage Sharing:

* **Goal**: Allow users to load model weights once into CPU memory (via from_file) and share/map/register that storage directly to the GPU that allows GPU access.
* **Benefit**: 
  * Fast Cold Start: Eliminates the GPU memcpy latency during model initialization.
  * Reduced Memory Footprint: The data exists only once in physical memory, effectively halving the RAM usage during the loading process compared to the standard pipeline.

### 1. `UsmAllocator` (CPU Side)

Implemented in `aten/src/ATen/UsmAllocator.cpp`. To support zero-copy sharing, the memory must meet specific alignment requirements and ideally bypass the OS page cache for speed.

* **Hugepage:** To reduce the GPU TLB stress. We choose to use `THP` on linux to get the hugepages transparently. But we haven't found the interface to apply for a large page in user mode yet on MacOS with Apple Silicon, so there may be some loss in memory access performance. 
* **Direct I/O:** Uses `O_DIRECT` (Linux) or `F_NOCACHE` (macOS) to read files directly into user memory, reducing system overhead.
* **Alignment:** Ensures allocations are page-aligned.

### 2. Zero-Copy `usm_share_` Interface

A new method `usm_share_(device)` is added to `Storage`. It takes an existing CPU storage and creates a device-mapped alias.

* **CUDA (Integrated, NVIDIA and AMD):**
  * Uses `cudaHostRegister` to register the existing CPU memory.
  * Uses `cudaHostGetDevicePointer` to obtain a mapped device pointer.
  * *Note:* Only enabled if `device_properties.integrated` is true.

* **MPS (Apple Silicon):**
  * Uses `[device newBufferWithBytesNoCopy: ...]` to wrap the existing CPU pointer into a `MTLBuffer`.

* **XPU (Intel iGPU):**
  * Uses `zeMemAllocShared` with `ZEX_HOST_MEM_ALLOC_FLAG_USE_HOST_PTR` to register the existing CPU buffer. As the support for Intel XPU is in the third party repo, a custom version is currently provided for testing. (https://github.com/nagic0/pytorch/tree/dev/usm_storage_intel and https://github.com/nagic0/torch-xpu-ops/tree/dev/torch_usm_l0)

### 3. API Changes

* **`torch.UntypedStorage.from_file`**: Added `usm=True` argument.
* **`storage.usm_share_(device)`**: New API to bridge CPU storage to Device storage.

## Usage & Performance Benefit

This serves as an optimization for model loaders (like `safetensors` or `torch.load` custom implementations).

For detailed reproduction steps, correctness verification, and full benchmark scripts, please refer to the demo repository: **[nagic0/torch-usm-demo](https://github.com/nagic0/torch-usm-demo)**.

### 1. Usage Example

```python
import torch

# 1. Efficient Loading (Direct I/O, Aligned)
# No OS Page Cache overhead, ready for sharing.
filename = "large_llm_weight.bin"
cpu_storage = torch.UntypedStorage.from_file(filename, size=..., usm=True)

# 2. Zero-Copy Mapping
# Instead of cpu_tensor.to("cuda"), which triggers a copy:
if torch.cuda.is_available():
    # Instantly maps the CPU pointer to a GPU pointer
    gpu_storage = cpu_storage.usm_share_(torch.device("cuda:0")) 
    
    # Create tensor wrapper
    gpu_tensor = torch.tensor([], device="cuda:0").set_(gpu_storage)

# Result: GPU tensor is ready immediately without memory bandwidth consumption.

```

### 2. Storage Micro-benchmark (SSD to GPU)

We measured the time taken to load a large binary file from an NVMe SSD into a CPU storage and then make it accessible on the GPU.

* **Copy Path:** `from_file(usm=False)` -> `storage.to(device)` (incurs CPU-to-GPU memcpy).
* **Share Path (USM):** `from_file(usm=True)` -> `storage.usm_share_(device)` (zero-copy mapping).

| Platform | Device | File Size | Copy Path | Share Path | Time Reduction |
| --- | --- | --- | --- | --- | --- |
| **NVIDIA Jetson AGX Orin** | CUDA | 5000 MB | 5.21s | 3.41s | **34.5%** |
| **AMD Strix Point** | CUDA (HIP) | 4000 MB | 3.22s | 2.01s | **37.8%** |
| **Intel Arrow Lake** | XPU (Level Zero) | 4000 MB | 2.29s | 0.86s | **62.2%** |
| **Apple M4 Pro** | MPS (Metal) | 4000 MB | 1.86s | 0.80s | **56.7%** |

### 3. LLM Cold Start Benchmark (`safetensors`)

We integrated this USM path into custom builds of `safetensors` and `transformers` to measure the real-world impact on loading Large Language Models (LLMs). This test includes the overhead of JSON parsing and tensor creation, strictly isolating the weight loading phase.

> **Note:** To reproduce these results, custom versions of `safetensors` and `transformers` that utilize the `usm=True` flag are required.

| Platform | Model | Copy Path | Share Path | Time Reduction |
| --- | --- | --- | --- | --- |
| **NVIDIA Jetson AGX Orin** | Qwen/Qwen3-8B | 19.10s | 10.93s | **43%** |
| **AMD Strix Point** | Qwen/Qwen3-8B | 12.73s | 8.25s | **35%** |
| **Intel Arrow Lake** | Gemma-3-27B | 38.41s | 11.20s | **71%** |
| **Apple M4 Pro** | Qwen/Qwen3-8B | 8.58s | 1.77s | **79%** (Has performance impact on decode stage, due to the lack of support for hugepages.) |

## Testing

* **Unit Tests:** Added a basic `test/test_usm_storage.py` covering lifecycle, ref-counting, and data correctness on both CUDA and MPS. 

## Future Work

This PR sets the foundation (`UsmAllocator` and `usm_share` API). Future work could promote the implementation of the `usm_share` interface in [intel/torch-xpu-ops](https://github.com/intel/torch-xpu-ops) and explore extending this to the default allocator or handling dynamic tensor allocations, but those are significantly more complex and outside the scope of this PR.

## Credits

This optimization is contributed by the Institute of Parallel and Distributed Systems (IPADS).

--------------

This is my first contribution to a large-scale open-source project, so please bear with me if I am unfamiliar with some of the workflows.

The initial development and testing were based on the PyTorch 2.9 release. However, due to merge conflicts, I have cherry-picked the changes onto the current main branch. ~~I am currently working on rebuilding and testing to ensure correctness and performance behavior are consistent with the original implementation.~~ (Done)

### CODE REVIEW

**Comment by nagic0** [comment]
## Rebuild and Check

* **AMD APU with CUDA device:** success build and test
* **NVIDIA Jetson with CUDA device:** need to fix the bug of building `ActivationPreluKernel.cu`, waiting for test
* **Apple Silicon with MPS device:** success build and test
* **Intel iGPU with XPU device:** success build and test

### Details about NVIDIA

#### Build error

The CUDA build for ActivationPreluKernel.cu was failing with errors in [Loops.cuh](https://github.com/nagic0/pytorch/blob/dev/usm_storage/aten/src/ATen/native/cuda/Loops.cuh#L67) at line 67. The error was:

```
error: calling a __host__ function std::apply from a __device__ function is not allowed
```

The code was using std::apply (a standard library host function) inside device code (CUDA kernel). NVCC strictly prohibits calling host functions from device code. This creates a device/host mismatch that the CUDA compiler rejects.

Replace `std::apply` with `c10::guts::apply` from [C++17.h](https://github.com/nagic0/pytorch/blob/dev/usm_storage/c10/util/C%2B%2B17.h#L56). The `c10::guts::apply` is marked with `__host__ __device__` attributes, making it compatible with both host and device code contexts. This function already works correctly in AMD builds.

### Details about Intel

#### Kernel AOT

If you are using Arrow Lake or Meteor Lake devices, remember to check if the [compute runtime](https://github.com/intel/compute-runtime) version is `25.48.36300.8` or newer. The old version of the driver has a bug that causes the kernel AOT of the pytorch operator to fail, triggering slow JIT compilation when executing pytorch code.

**Comment by lostdisc** [comment]
I'm still interested in this, or something like it.
